scripts/dts/gen_defines.py

```diff
@@ -253,23 +253,69 @@ def relativize(path):
 def write_regs(node):
     # Writes address/size output for the registers in the node's 'reg' property
 
-    def write_reg(reg, base_ident, val):
-        # Drop '_0' from the identifier if there's a single register, for
-        # backwards compatibility
-        if len(reg.node.regs) > 1:
-            ident = f"{base_ident}_{reg.node.regs.index(reg)}"
-        else:
-            ident = base_ident
-
-        out_node(node, ident, val,
-                 # Name alias from 'reg-names = ...'
-                 f"{str2ident(reg.name)}_{base_ident}" if reg.name else None)
+    if not node.regs:
+        return
 
-    for reg in node.regs:
-        write_reg(reg, "BASE_ADDRESS", hex(reg.addr))
-        if reg.size:
-            write_reg(reg, "SIZE", reg.size)
+    primary_addrs = {}
+    primary_sizes = {}
+
+    def write_regs_for_ident(ident):
+        # Write BASE_ADDRESS and SIZE macros for a given identifier
+        # 'ident'. If we have already generated primary address and
+        # size macros and saved in them in primary_addrs and
+        # primary_sizes, we just reuse those. Otherwise (i.e. the
+        # first time this is called), they are generated from the
+        # actual reg.addr and reg.size attributes, and the names of
+        # the primary macros are saved.
+
+        for reg_i, reg in enumerate(node.regs):
+            idx = f"_{reg_i}"
+            name = f"{str2ident(reg.name)}_" if reg.name else ""
+            addr = primary_addrs.get(reg_i)
+            size = primary_sizes.get(reg_i)
```

Comment:
Maybe this could be refactored to remove the `addr = None ... if addr is None:` indirection:

```python
if reg_i not in primary_addr:
    primary_addrs[reg_i] = prim_addr
    primary_sizes[reg_i] = prim_size

reg = primary_addrs[reg_i]
addr = primary_sizes[reg_i]

write_reg(...)
```

Could maybe merge `primary_addrs` and `primary_sizes` too, since they're always accessed together, and maybe use a list, since you never get any "holes" in the indices (though dict is fine if it turns out clearer).